src/metrics/metrics_provider.cpp

```diff
@@ -21,27 +21,103 @@
 
-#include <QByteArray>
 #include <QDateTime>
 #include <QEventLoop>
-#include <QJsonArray>
 #include <QJsonDocument>
 #include <QJsonObject>
 #include <QNetworkReply>
 #include <QNetworkRequest>
 #include <QRegExp>
-#include <QString>
 
 namespace mp = multipass;
 namespace mpl = multipass::logging;
 
+using namespace std::literals::chrono_literals;
+
 namespace
 {
 constexpr auto category = "metrics";
+
+void post_request(const QUrl& metrics_url, const QByteArray& body)
+{
+    QNetworkAccessManager manager;
+    QNetworkRequest request{metrics_url};
+    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+    request.setHeader(QNetworkRequest::ContentLengthHeader, body.size());
+    request.setHeader(QNetworkRequest::UserAgentHeader, "multipassd/1.0");
+
+    QEventLoop event_loop;
+    std::unique_ptr<QNetworkReply> reply;
+
+    // For unit testing
+    if (metrics_url.isLocalFile())
+        reply = std::unique_ptr<QNetworkReply>(manager.put(request, body));
+    else
+        reply = std::unique_ptr<QNetworkReply>(manager.post(request, body));
+
+    QObject::connect(reply.get(), &QNetworkReply::finished, &event_loop, &QEventLoop::quit);
+
+    event_loop.exec();
+
+    auto buff = reply->readAll();
+
+    if (reply->error() == QNetworkReply::ProtocolInvalidOperationError)
+    {
+        auto error_msg = QJsonDocument::fromJson(buff).object();
+        mpl::log(mpl::Level::error, category,
+                 fmt::format("Metrics error: {} - {}", error_msg["code"].toString().toStdString(),
+                             error_msg["message"].toString().toStdString()));
+    }
+    else if (reply->error() != QNetworkReply::NoError)
+    {
+        QObject::disconnect(reply.get(), 0, 0, 0);
+        throw std::runtime_error(fmt::format("Metrics error: {}", reply->errorString().toStdString()));
+    }
+}
 }
 
-mp::MetricsProvider::MetricsProvider(const QUrl& metrics_url, const QString& unique_id)
-    : metrics_url{metrics_url}, unique_id{unique_id}
+mp::MetricsProvider::MetricsProvider(const QUrl& url, const QString& unique_id)
+    : metrics_url{url}, unique_id{unique_id}, metrics_sender{[this] {
+          std::unique_lock<std::mutex> lck(metrics_mutex, std::defer_lock);
+          while (true)
+          {
+              while (!lck.try_lock())
+                  std::this_thread::sleep_for(1ms);
+
+              if (metric_batches.empty())
+              {
+                  metrics_cv.wait(lck);
+              }
+
+              // Null element signals thread to exit
+              if (metric_batches.first().isNull())
+              {
+                  lck.unlock();
+                  return;
+              }
+
+              auto body = QJsonDocument(metric_batches).toJson(QJsonDocument::Compact);
+              lck.unlock();
+
+              try
+              {
+                  post_request(metrics_url, body);
+                  lck.lock();
+                  if (metric_batches.first().isNull())
```

Comment:
Once the metrics are sent why are they being checked again? I think an explicit exit flag would be clearer.

Reply:
A null QJsonValue is the indicator that the thread needs to exit, so after posting the metrics, the thread should check for the NULLness.  Sure a flag would be fine too, but I was just avoiding cluttering up the class a bit, but it would probably be clearer and also would allow for the explicit thread exit.  I'll just use an atomic bool for this.